File: src/main/java/net/tonimatasdev/GamePanel.java
package net.tonimatasdev;

import javax.swing.*;
import java.awt.*;

public class GamePanel extends JPanel implements Runnable {

    // SCREEN SETTINGS
    final  int originalTileSize = 16; // 16x16 tile
    final int scale = 3;

    final int tileSize = originalTileSize * scale; // 48x48 tile
    final int maxScreenCol = 16;
    final int maxScreenRow = 12;
    final int screenWidth = tileSize * maxScreenCol;
    final int screenHeight = tileSize * maxScreenRow;

    int FPS = 60;
    KeyHandler keyHandler = new KeyHandler();
    Thread gameThread;

    int playerX = 100;
    int playerY = 100;
    int playerSpeed = 4;

    public GamePanel() {
        this.setPreferredSize(new Dimension(screenWidth, screenHeight));
        this.setBackground(Color.BLACK);
        this.setDoubleBuffered(true);
        this.addKeyListener(keyHandler);
        this.setFocusable(true);
    }

    public void startGameThread() {
        gameThread = new Thread(this);
        gameThread.start();
    }

    @Override
    public void run() {
        double drawInterval = 1000000000d / FPS;
        double nextDrawTime = System.nanoTime() + drawInterval;

        while (gameThread != null) {

            update();

            repaint();

            try {
                double reamingTime = nextDrawTime - System.nanoTime();
                reamingTime = reamingTime / 1000000;

                if (reamingTime < 0) {
                    reamingTime = 0;
                }

                Thread.sleep((long) reamingTime);

                nextDrawTime += drawInterval;
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public void update() {
        if (keyHandler.upPressed) {
            playerY -= playerSpeed;
        } else if (keyHandler.downPressed) {
            playerY += playerSpeed;
        } else if (keyHandler.leftPressed) {
            playerX -= playerSpeed;
        } else if (keyHandler.rightPressed) {
            playerX += playerSpeed;
        }
    }

    public void paintComponent(Graphics graphics) {
        super.paintComponent(graphics);

        Graphics2D graphics2D = (Graphics2D) graphics;

        graphics2D.setColor(Color.WHITE);
        graphics2D.fillRect(playerX, playerY, tileSize, tileSize);
        graphics2D.dispose();
    }
}

File: src/main/kotlin/dev/tonimatas/game/GamePanel.kt
package dev.tonimatas.game

import dev.tonimatas.game.entity.Player
import dev.tonimatas.game.tile.TileManager
import java.awt.Color
import java.awt.Dimension
import java.awt.Graphics
import java.awt.Graphics2D
import javax.swing.JPanel

class GamePanel() : JPanel(), Runnable {
    // SCREEN SETTINGS
    val originalTileSize: Int = 16 // 16x16 tile
    val scale: Int = 3

    val tileSize: Int = originalTileSize * scale // 48x48 tile
    val maxScreenColumn: Int = 16
    val maxScreenRow: Int = 12
    val screenWidth: Int = tileSize * maxScreenColumn
    val screenHeight: Int = tileSize * maxScreenRow

    // FPS
    var FPS: Int = 60
    
    val tileManager = TileManager(this)
    val keyHandler = KeyHandler()
    var gameThread: Thread? = null
    val player = Player(this, this.keyHandler)

    init {
        this.preferredSize = Dimension(screenWidth, screenHeight)
        this.background = Color.BLACK
        this.isDoubleBuffered = true
        this.addKeyListener(keyHandler)
        this.isFocusable = true
    }


    fun startGameThread() {
        gameThread = Thread(this)
        gameThread!!.start()
    }

    override fun run() {
        val drawInterval = 1000000000.0 / FPS
        var nextDrawTime = System.nanoTime() + drawInterval

        while (gameThread != null) {
            update()
            repaint()
            
            try {
                var reamingTime = nextDrawTime - System.nanoTime()
                reamingTime = reamingTime / 1000000

                if (reamingTime < 0) {
                    reamingTime = 0.0
                }

                Thread.sleep(reamingTime.toLong())

                nextDrawTime += drawInterval
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
        }
    }

    private fun update() {
        player.update()
    }

    public override fun paintComponent(graphics: Graphics) {
        super.paintComponent(graphics)

        val graphics2D = graphics as Graphics2D

        tileManager.draw(graphics2D)
        
        player.draw(graphics2D)
    }
}